src/liballoc/alloc.rs

```diff
@@ -181,33 +191,81 @@ unsafe impl AllocRef for Global {
     }
 
     #[inline]
-    unsafe fn realloc(
+    unsafe fn grow(
         &mut self,
         ptr: NonNull<u8>,
         layout: Layout,
         new_size: usize,
+        placement: ReallocPlacement,
+        init: AllocInit,
     ) -> Result<(NonNull<u8>, usize), AllocErr> {
-        match (layout.size(), new_size) {
-            (0, 0) => Ok((layout.dangling(), 0)),
-            (0, _) => self.alloc(Layout::from_size_align_unchecked(new_size, layout.align())),
-            (_, 0) => {
-                self.dealloc(ptr, layout);
-                Ok((layout.dangling(), 0))
+        let old_size = layout.size();
+        debug_assert!(
+            new_size >= old_size,
+            "`new_size` must be greater than or equal to `layout.size()`"
+        );
+
+        // `new_size == 0` implies `old_size == 0`
+        if old_size == new_size || new_size == 0 {
```

Comment:
```suggestion
        if old_size == new_size {
```
If `new_size == 0`, then since `new_size >= old_size` and they are both unsigned, `old_size == 0` as mentioned in the comment above. Therefore, `old_size == new_size`.